CONTRIBUTING.md

```diff
@@ -524,6 +490,40 @@ into any of those categories, use your best judgement.
 * Use `HasX` for protocols that have readable and/or writable attributes
   or getter/setter methods (e.g. `HasItems`, `HasFileno`).
 
+### Incomplete stubs
+
+We accept partially annotated stubs, especially for larger packages.
+These need to follow the following guidelines:
+
+* A package must include all public submodules, all public classes,
+  functions and methods, and all other public fields.
+* Included functions and methods must list all arguments, but the argument
+  types and the return type can be left unannotated or only be partially
+  annotated.
+* A type can be partially annotated by using the `_typeshed.Incomplete`
+  marker. A field can be left unannotated by using `Incomplete` as well.
+  Do not use `Incomplete` unless it's syntactically required. See below for
+  an example.
+
+Example of an unannotated field and a partially annotated function:
+
+```python
+from _typeshed import Incomplete
+
+pub_field1: Incomplete  # unannotated
+pub_field2: list[Incomplete]  # partially annotated
```

Comment:
Something I've noticed for a short while now. And may require further discussion. Is that typeshed is "lying" and hiding "partially unknown types" to strict pyright users.

Reply:
Possible, maybe @erictraut can shed some light on that.

Reply:
I think it's a good idea to use `Incomplete` as an alias for `Any` (in cases where it is syntactically required) to indicate that someone hasn't yet made the investment to provide a correct type annotation. That makes it clear that it's a "work in progress". Contributors who want to improve typeshed can then focus their attention on eliminating `Incomplete`. 

The result for strict pyright users isn't any worse than if `Any` were used in these locations.

Reply:
Wouldn't `pub_field2: list` allow to warn users about using unknown types through `reportUnknownMemberType` and `reportUnknownVariableType` ?

Reply:
Yes, that's correct. Since `[Incomplete]` isn't syntactically required, the guidance would seem to indicate that it should be omitted in this particular case.

Reply:
Good point, I will update the wording to require `Incomplete` in these cases.

Reply:
I think @Avasam is arguing that the wording is fine but the code sample should be changed — that `pub_field2` should use `list` rather than `list[Incomplete]`. A type annotation of `list` is more useful for users of pyright who are using strict-mode checks because it will be evaluated as `list[Unknown]` which will result in an error if it is used. By contrast, `list[Incomplete]` will be interpreted as `list[Any]` which will not result in any strict-mode errors.

Reply:
We have handled this differently in typeshed so far. Using bare `list` would trigger "missing generic argument" warnings if I remember correctly. And I think this is a valid warning that we should keep.

Reply:
Exactly what Eric Traut said.
It would require removing `"reportMissingTypeArgument": "error"` from the non-stricter configuration though. Which can be done separately from this PR (and then update the doc). I just wanted to bring attention to this.

Reply:
Or you could use `# pyright: ignore[reportMissingTypeArgument]` to suppress errors on individual lines.

Reply:
I don't like either approach. I think we should continue using `list[Incomplete]`.

Reply:
I don't have a strong opinion. It sounds like there's an established precedent in typeshed already, so it makes sense to stick with it.

@Avasam, if you're concerned about specific "holes" created by the use of `Incomplete`, you can plug those holes by contributing improvements that replace those instances of `Incomplete` with the proper type.

Reply:
One thing that would really help is a proper `typing.Unknown` type in Python. But that gets too off-topic for this PR.

Reply:
Also, since I didn't give a proper justification for my opinion:

* `reportMissingTypeArgument` is a useful warning, not only in completed stubs. Disabling it would be a disservice.
* Using per-instance warning suppression would only work for pyright, but I think mypy warns about it as well. Also, `list[Incomplete]` is more concise and it's immediately obvious that the annotation is incomplete. It's also greppable.

Reply:
I had a vague recollection that pyright was meant to treat the `_typeshed.Incomplete` alias like `Unknown`. Sounds like that didn't happen, but it would solve the issue here.

Reply:
If pyright did treat `Incomplete` like `Unknown`, presumably we'd get a warning from pyright about an unknown (or partially unknown) type every time we did `x: Incomplete` or `y: list[Incomplete]`, so we'd have to `pyright: ignore` it every time we did something like that in a stub with our stricter pyright settings enabled.

Reply:
@AlexWaygood But then you'd disable the rule entirely in typeshed because it wouldn't mean anything for typeshed stubs (`list` would be equivalent to `list[Incomplete]`, so no point in raising an error).

Also, relevant question I had asked on pyright discussion forum and Eric's answer: https://github.com/microsoft/pyright/discussions/4471

> Incomplete isn't an official symbol in the type system or in the Python runtime, and it's not documented in any PEPs. I'd therefore be reluctant to give it any special-case meaning in pyright. It's defined in typeshed.pyi as a type alias for Any, so that's how pyright will treat it.

Speaking of which, is there a proposition or something somewhere that I could upvote for official `Unknown`/`Incomplete` support? I heard that idea being thrown around a handful of times.

Reply:
> Speaking of which, is there a proposition or something somewhere that I could upvote for official `Unknown`/`Incomplete` support? I heard that idea being thrown around a handful of times.

I don't know of such a proposal, but I'd also upvote it if there were one. I think it's a very useful concept, and I've often thought it was a shame mypy didn't have something similar.